controllers/osccluster_publicip_controller_unit_test.go

```diff
@@ -0,0 +1,700 @@
+			if err != nil {
+				assert.Equal(t, pitc.expCheckPublicIpFormatParametersErr, err, "checkPublicIpFormatParameters() should return the same error")
+			} else {
+				assert.Nil(t, pitc.expCheckPublicIpFormatParametersErr)
+			}
+			t.Logf("find publicIpName %s\n", publicIpName)
+		})
+	}
+}
+
+// TestCheckPublicIpOscAssociateResourceName has several tests to cover the code of the function checkPublicIpOscAssociateResourceName
+func TestCheckPublicIpOscAssociateResourceName(t *testing.T) {
+	publicIpTestCases := []struct {
+		name                                        string
+		spec                                        infrastructurev1beta1.OscClusterSpec
+		expCheckPublicIpOscAssociateResourceNameErr error
+	}{
+		{
+			name: "check natservice association with publicIp",
+			spec: infrastructurev1beta1.OscClusterSpec{
+				Network: infrastructurev1beta1.OscNetwork{
+					Net: infrastructurev1beta1.OscNet{
+						Name:    "test-net",
+						IpRange: "10.0.0.0/16",
+					},
+					NatService: infrastructurev1beta1.OscNatService{
+						Name:         "test-natservice",
+						PublicIpName: "test-publicip",
+						SubnetName:   "test-subnet",
+					},
+					PublicIps: []*infrastructurev1beta1.OscPublicIp{
+						{
+							Name: "test-publicip",
+						},
+					},
+				},
+			},
+			expCheckPublicIpOscAssociateResourceNameErr: nil,
+		},
+		{
+			name: "check natService association with bad publicIp",
+			spec: infrastructurev1beta1.OscClusterSpec{
+				Network: infrastructurev1beta1.OscNetwork{
+					Net: infrastructurev1beta1.OscNet{
+						Name:    "test-net",
+						IpRange: "10.0.0.0/16",
+					},
+					NatService: infrastructurev1beta1.OscNatService{
+						Name:         "test-natservice",
+						PublicIpName: "test-publicip-test",
+						SubnetName:   "test-subnet",
+					},
+					PublicIps: []*infrastructurev1beta1.OscPublicIp{
+						{
+							Name: "test-publicip",
+						},
+					},
+				},
+			},
+			expCheckPublicIpOscAssociateResourceNameErr: fmt.Errorf("publicIp test-publicip-test-uid does not exist in natService "),
+		},
+	}
+	for _, pitc := range publicIpTestCases {
+		t.Run(pitc.name, func(t *testing.T) {
+			clusterScope := Setup(t, pitc.name, pitc.spec)
+			err := checkPublicIpOscAssociateResourceName(clusterScope)
+			if err != nil {
+				assert.Equal(t, pitc.expCheckPublicIpOscAssociateResourceNameErr, err, "checkPublicIpOscAssociateResourceName() should return the same error")
+			} else {
+				assert.Nil(t, pitc.expCheckPublicIpOscAssociateResourceNameErr)
+			}
+		})
+	}
+}
+
+// TestCheckPublicIpOscDuplicateName has several tests to cover the code of the function checkPublicIpOscDuplicateName
+func TestCheckPublicIpOscDuplicateName(t *testing.T) {
+	publicIpTestCases := []struct {
+		name                                string
+		spec                                infrastructurev1beta1.OscClusterSpec
+		expCheckPublicIpOscDuplicateNameErr error
+	}{
+		{
+			name:                                "get separate Name",
```

Comment:
What do you mean by that  ("Get separate Name") ?

Reply:
Different maybe ?

Reply:
I change from separate to distinct.